### PR TITLE
Allow admin to associate missing LFS objects for repositories

### PR DESCRIPTION
**Pull request by wxiaoguang**
# Background

There are some flaws in Gitea's LFS module.

1. Adopted repositories don't contain LFS objects
2. Missing LFS objects (which do not exist in lfs_meta_object table) can not be associated to a repository
3. Forked/rsynced repositories might not serve LFS objects correctly

Some related issues:

* https://github.com/go-gitea/gitea/issues/18137
* https://github.com/go-gitea/gitea/issues/17715
* https://github.com/go-gitea/gitea/issues/17207

# Solution (partially)

This PR doesn't resolve all these issues.

This PR reworked the `Find pointer files` feature in Settings -> LFS page.

When a LFS object is missing from database, admin can associate it to the repository by clicking the `Associate` button.

This PR did help a user to recover their LFS repositories successfully.

I agree that this PR is not perfect, because the LFS module itself should be improved too, but I think this feature is nice to have.

image

### CODE REVIEW

File: routers/web/repo/lfs.go [hunk 1]
```diff
@@ -461,22 +462,29 @@ func LFSPointerFiles(ctx *context.Context) {
...
 					}
-				} else {
-					result.Accessible = true
+					if !result.Associatable {
+						associated, err := models.LFSObjectIsAssociated(pointerBlob.Oid)
+						if err != nil {
+							return err
+						}
+						result.Associatable = !associated
+					}
```
Comment (zeripath) [inline routers/web/repo/lfs.go:475]: I don't understand why this is necessary

File: models/lfs.go [hunk 2]
```diff
@@ -162,23 +169,39 @@ func LFSAutoAssociate(metas []*LFSMetaObject, user *user_model.User, repoID int6
...
+		// admin can associate any LFS object to any repository, and we do not care about errors (eg: duplicated unique key),
+		// even if error occurs, it won't hurt users and won't make things worse
+		for i := range metas {
+			_, err = sess.Insert(&LFSMetaObject{
+				Pointer:      lfs.Pointer{Oid: metas[i].Oid, Size: metas[i].Size},
+				RepositoryID: repoID,
+			})
+			if err != nil {
+				log.Warn("failed to insert LFS meta object into database, err=%v", err)
+			}
```
Comment (zeripath) [inline models/lfs.go:202]: If the content store does not contain the LFS meta ID won't this result in dangling pointers and corruption of the content store? Shouldn't we be checking that the physical store contains something with this ID?

File: routers/web/repo/lfs.go (hunk 1 above)
Comment (wxiaoguang) [reply routers/web/repo/lfs.go:475]: What do you mean about `this`? 

If `this` means this PR: it just adds a feature to help users to fix their broken LFS (eg: database was lost, table was truncated)

If `this` means the code block: I think it's better to make sure that there is no record in database. If the LFS oid already exists in database table, then the oid should have been associated to other repositories, then we need the `LFSObjectAccessible` above to check if the user can access it.

I am still not sure what's the best way to do the associating, so suggestions are welcome.

File: models/lfs.go (hunk 2 above)
Comment (wxiaoguang) [reply models/lfs.go:202]: The `metas` comes from the `pointerResult` slice in `routers/web/repo/lfs.go`, it has been check by `result.Exists, err = contentStore.Exists(pointerBlob.Pointer)`.

The old logic is: get a valid pointerResult slice (checked by content store) -> generate a form with `oid size` list -> submit form -> parse the `oid size` list to `metas` -> insert into database. I didn't change these code (although that's quite hacky).

**Review by zeripath — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
trigger bot

**Pull request merged**